README.md
# taipy2 README

This is the README for your extension "taipy2". 

## Features
Autocomplete and suggestion of code, i.e. code completion with custom local hosted model API

## Requirements

A server API must be ready. 
Update this line 129 in the extension.js with your own API
    const apiUrl = 'https://aee0-103-253-89-37.ngrok-free.app/api/generate';

Instruction for server installation as follows:
- Build a local version of quantized, run the following command:

_"ct2-transformers-converter --model thanhnew2001/starcoder-7b-taipy5 --output_dir taipy5-ct2 --force --copy_files tokenizer.json tokenizer_config.json  special_tokens_map.json .gitattributes --quantization int8_float16 --trust_remote_code"_

- You will need at least 8GB VRAM to run this API
Run it: **python server_infer_ngrok_taipy.py**
Copy the ngrok generated API and replace the line 129 in the extension.

## Extension Settings

Nothing special needed

## Known Issues

Model needs more data for fine-tuning

## Release Notes
This is for testing and it works quite well

### 1.0.0

Initial release of ...

### 1.0.1

Fixed issue #.

### 1.1.0

Added features X, Y, and Z.

---


## For more information

* [Visual Studio Code's Markdown Support](http://code.visualstudio.com/docs/languages/markdown)
* [Markdown Syntax Reference](https://help.github.com/articles/markdown-basics/)

**Enjoy!**


CHANGELOG.md
# Change Log

All notable changes to the "taipy2" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

server_infer_ngrok_taipy.py
import os
import threading

from flask import Flask
from pyngrok import ngrok
from hf_hub_ctranslate2 import GeneratorCT2fromHfHub

from flask import request, jsonify

model_name = "taipy5-ct2" # note this is local folder model, the model uploaded to huggingface did not response correctly
#model_name = "michaelfeil/ct2fast-starchat-alpha"
#model_name = "michaelfeil/ct2fast-starchat-beta"
model = GeneratorCT2fromHfHub(
        # load in int8 on CUDA
        model_name_or_path=model_name,
        device="cuda",
        compute_type="int8_float16",
        # tokenizer=AutoTokenizer.from_pretrained("{ORG}/{NAME}")
)

def generate_text_batch(prompt_texts, max_length=64):
    outputs = model.generate(prompt_texts, max_length=max_length, include_prompt_in_result=False)
    return outputs

app = Flask(__name__)
port = "5000"

# Open a ngrok tunnel to the HTTP server
public_url = ngrok.connect(port).public_url
print(" * ngrok tunnel \"{}\" -> \"http://127.0.0.1:{}\"".format(public_url, port))

# Update any base URLs to use the public ngrok URL
app.config["BASE_URL"] = public_url

# ... Update inbound traffic via APIs to use the public-facing ngrok URL
# Define Flask routes
@app.route("/")
def index():
    return "Hello from Colab!"

@app.route("/api/generate", methods=["POST"])
def generate_code():
    try:
        # Get the JSON data from the request body
        data = request.get_json()
        
        # Extract 'inputs' and 'parameters' from the JSON data
        inputs = data.get('inputs', "")
        parameters = data.get('parameters', {})
        
        # Extract the 'max_new_tokens' parameter
        max_new_tokens = parameters.get('max_new_tokens', 64)

        # Call the generate_text_batch function with inputs and max_new_tokens
        generated_text = generate_text_batch([inputs], max_new_tokens)[0]

        return jsonify({
        "generated_text": generated_text,
        "status": 200
        })

    except Exception as e:
        return jsonify({"error": str(e)})


# Start the Flask server in a new thread
threading.Thread(target=app.run, kwargs={"use_reloader": False}).start()


extension.js
const vscode = require('vscode');
const axios = require('axios');

const CONTEXT_LENGTH = 32; // Default context length in tokens

let disposable;
let ghostTextDecorationType;
let timer; // Timer to trigger the API call after a pause
let lastCursorPosition; // Store the last cursor position for selected text
let userTyping = false; // Variable to track if the user is typing

// Maintain a list of active ghost text decorations
let activeGhostTextDecorations = [];

function activate(context) {
    console.log('Ghost Text Extension is active!');

    ghostTextDecorationType = vscode.window.createTextEditorDecorationType({
        after: {
            contentText: '', // Initialize contentText as empty
            color: '#888888', // Ghost text color
        },
    });

    disposable = vscode.window.onDidChangeTextEditorSelection(updateSelectedText);
    vscode.workspace.onDidChangeTextDocument(onTextDocumentChange);

    // Register the "insertGhostText" command to handle Tab key presses
    context.subscriptions.push(vscode.commands.registerCommand('extension.insertGhostText', insertGhostText));
}

function updateSelectedText(event) {
    const editor = vscode.window.activeTextEditor;
    if (!editor) return;

    // Get the selected text within the context length
    const selection = editor.selection;
    lastCursorPosition = editor.selection.active;

    const startPosition = new vscode.Position(
        Math.max(0, lastCursorPosition.line - CONTEXT_LENGTH),
        Math.max(0, lastCursorPosition.character - CONTEXT_LENGTH)
    );

    const selectedText = editor.document.getText(new vscode.Range(startPosition, lastCursorPosition));

    if (selectedText.length > CONTEXT_LENGTH * 2) {
        // Truncate the selected text if it's longer than double the context length
        const truncatedLength = selectedText.length - CONTEXT_LENGTH * 2;
        selectedText = selectedText.substring(truncatedLength);
    }

    if (!selection.isEmpty) {
        // User is typing, set the flag to true
        userTyping = true;

        // Clear the timer
        clearTimeout(timer);
    }
}

// ... (rest of the code remains the same)

// ... (rest of the code remains the same)


function onTextDocumentChange(event) {
    // Handle Tab key presses separately when a document is modified
    if (event.contentChanges[0]?.text === '\t') {
        insertGhostText();
    } 
    else {
        // Clear the active ghost text decorations
        activeGhostTextDecorations = [];
        // Update the editor's decorations to remove the ghost text
        const editor = vscode.window.activeTextEditor;
        if (editor) {
            editor.setDecorations(ghostTextDecorationType, activeGhostTextDecorations);
        }

        // User is not typing, trigger the API call
        if (!userTyping) {
            clearTimeout(timer);
            timer = setTimeout(() => {
                triggerAPICall(editor);
            }, 2000);
        }

        // Reset the userTyping flag
        userTyping = false;
    }
}


async function triggerAPICall(editor) {
    // Get the selected text before the cursor position
    const selectedText = editor.document.getText(new vscode.Range(new vscode.Position(0, 0), lastCursorPosition));

	console.log("triggerAPICall->"+selectedText)
    if (selectedText) {
        try {
            // Send the selected text to an LLM API and get the response.
            const ghostText = await sendTextToLLMAPI(selectedText);

            // Create a new decoration with updated contentText.
            const newDecoration = {
                range: new vscode.Range(lastCursorPosition, editor.selection.active),
                renderOptions: {
                    after: {
                        contentText: ghostText,
                        color: '#888888',
                    },
                },
            };

            // Add the new decoration to the list of active decorations
            activeGhostTextDecorations.push(newDecoration);

            // Apply all active decorations
            editor.setDecorations(ghostTextDecorationType, activeGhostTextDecorations);
        } catch (error) {
            console.error('Error fetching ghost text:', error);
        }
    }
}

async function sendTextToLLMAPI(text) {
    // Define the API endpoint and request data.
    const apiUrl = 'https://aee0-103-253-89-37.ngrok-free.app/api/generate';
    const requestData = {
        inputs: text,
        parameters: {
            max_new_tokens: 64,
        },
    };

    // console.log(requestData)
    // Send a POST request to the LLM API.
    try {
        const response = await axios.post(apiUrl, requestData);
        console.log(response.status)
        if (response.status === 200) {
            text = response.data.generated_text
            console.log(text)
            return text;
        } else {
            throw new Error(`LLM API request failed with status code: ${response.status}`);
        }
    } catch (error) {
        throw new Error(`Error sending text to LLM API: ${error.message}`);
    }
}

function insertGhostText() {
    const editor = vscode.window.activeTextEditor;

    if (!editor) return;

    // Check if there are active ghost text decorations
    if (activeGhostTextDecorations.length > 0) {
        const ghostText = activeGhostTextDecorations[0].renderOptions.after.contentText;

        // Insert the ghost text at the cursor position
        editor.edit(editBuilder => {
            editBuilder.insert(lastCursorPosition, ghostText);
        });

        // Remove the active ghost text decoration
        activeGhostTextDecorations = [];
        editor.setDecorations(ghostTextDecorationType, activeGhostTextDecorations);
    }
}

function deactivate() {
    disposable.dispose();
}

module.exports = {
    activate,
    deactivate,
};
